src/deploy/functions/services/auth.ts

```diff
@@ -0,0 +1,150 @@
+   */
+  validateTrigger(
+    endpoint: backend.Endpoint & backend.BlockingTriggered,
+    wantBackend: backend.Backend
+  ): void {
+    const blockingEndpoints = backend
+      .allEndpoints(wantBackend)
+      .filter((ep) => backend.isBlockingTriggered(ep)) as (backend.Endpoint &
+      backend.BlockingTriggered)[];
+    if (
+      blockingEndpoints.find(
+        (ep) =>
+          ep.blockingTrigger.eventType === endpoint.blockingTrigger.eventType &&
+          ep.id !== endpoint.id
+      )
+    ) {
+      throw new FirebaseError(
+        `Can only create at most one Auth Blocking Trigger for ${endpoint.blockingTrigger.eventType} events`
+      );
+    }
+  }
+
+  private configChanged(
+    newConfig: identityPlatform.BlockingFunctionsConfig,
+    config: identityPlatform.BlockingFunctionsConfig
+  ) {
+    if (
+      newConfig.triggers?.beforeCreate?.functionUri !==
+        config.triggers?.beforeCreate?.functionUri ||
+      newConfig.triggers?.beforeSignIn?.functionUri !== config.triggers?.beforeSignIn?.functionUri
+    ) {
+      return true;
+    }
+    if (
+      !!newConfig.forwardInboundCredentials?.accessToken !==
+        !!config.forwardInboundCredentials?.accessToken ||
+      !!newConfig.forwardInboundCredentials?.idToken !==
+        !!config.forwardInboundCredentials?.idToken ||
+      !!newConfig.forwardInboundCredentials?.refreshToken !==
+        !!config.forwardInboundCredentials?.refreshToken
+    ) {
+      return true;
+    }
+    return false;
+  }
+
+  private async registerTriggerLocked(
+    endpoint: backend.Endpoint & backend.BlockingTriggered
+  ): Promise<void> {
+    const newBlockingConfig = await identityPlatform.getBlockingFunctionsConfig(endpoint.project);
+    const oldBlockingConfig = cloneDeep(newBlockingConfig);
+
+    if (endpoint.blockingTrigger.eventType === events.v1.BEFORE_CREATE_EVENT) {
+      newBlockingConfig.triggers = {
+        ...newBlockingConfig.triggers,
+        beforeCreate: {
+          functionUri: endpoint.uri!,
+        },
+      };
+    } else {
+      newBlockingConfig.triggers = {
+        ...newBlockingConfig.triggers,
+        beforeSignIn: {
+          functionUri: endpoint.uri!,
+        },
+      };
+    }
+
+    newBlockingConfig.forwardInboundCredentials = {
+      idToken: endpoint.blockingTrigger.options?.idToken || false,
+      accessToken: endpoint.blockingTrigger.options?.accessToken || false,
+      refreshToken: endpoint.blockingTrigger.options?.refreshToken || false,
+    };
+
+    if (!this.configChanged(newBlockingConfig, oldBlockingConfig)) {
+      return;
+    }
+
+    await identityPlatform.setBlockingFunctionsConfig(endpoint.project, newBlockingConfig);
+  }
+
+  /**
+   * Registers the auth blocking trigger to identity platform.
+   * @param ep the blocking endpoint
+   */
+  registerTrigger(ep: backend.Endpoint & backend.BlockingTriggered): Promise<void> {
+    this.triggerQueue = this.triggerQueue.then(() => this.registerTriggerLocked(ep));
```

Comment:
I pushed Cole to this implementation. The problem is that we update functions concurrently and it's possible that we update the beforeCreate function at the same time as beforeSignIn. Since the process of updating involves reading a whole config, changing a field, and saving a whole config, it's possible that two racing updates will read the same original config, change one field, and only one write will win.

By using a promise chain we guarantee that the racing updates will happen sequentially. Think of it as running inside a mutex.